File: python/deepstrike/runtime/replay_fixture.py
"""Fixture helpers for ReplayProvider.

Python port of node/src/runtime/replay-fixture.ts. Walks `llm_completed` events from a
session log and returns the ordered list of assistant Messages.

Accepts both wire shapes the SDK uses interchangeably:
- camelCase in-memory: `toolCalls`, `tokenCount`
- snake_case on-disk:  `tool_calls`, `token_count`
"""
from __future__ import annotations

from typing import Any, Iterable

from deepstrike._kernel import Message  # type: ignore


def extract_recorded_messages(events: Iterable[Any]) -> list[Message]:
    """Walk session events (either wrapped `{seq, event}` or bare events) into Message[]."""
    import json

    out: list[Message] = []
    for entry in events:
        event = entry.get("event") if isinstance(entry, dict) and "event" in entry else entry
        kind = event.get("kind") if isinstance(event, dict) else getattr(event, "kind", None)
        if kind != "llm_completed":
            continue

        def _g(k: str) -> Any:
            if isinstance(event, dict):
                return event.get(k)
            return getattr(event, k, None)

        content = _g("content") or ""
        # camelCase first, snake_case fallback
        raw_tc = _g("toolCalls") or _g("tool_calls") or []
        normalized_tc: list[dict] = []
        for tc in raw_tc:
            name = tc.get("name") if isinstance(tc, dict) else getattr(tc, "name", "")
            tc_id = tc.get("id") if isinstance(tc, dict) else getattr(tc, "id", "")
            args = tc.get("arguments") if isinstance(tc, dict) else getattr(tc, "arguments", "")
            if not isinstance(args, str):
                args = json.dumps(args)
            normalized_tc.append({"id": tc_id, "name": name, "arguments": args})
        token_count = _g("tokenCount") or _g("token_count")

        msg: dict[str, Any] = {"role": "assistant", "content": content}
        if normalized_tc:
            msg["toolCalls"] = normalized_tc
        if token_count is not None:
            msg["tokenCount"] = token_count
        out.append(msg)  # type: ignore[arg-type]
    return out

File: wasm/src/runtime/replay-fixture.ts
/**
 * Fixture helpers for `ReplayProvider`.
 *
 * The canonical persistence shape for a recorded run is a `SessionLog` of `llm_completed` events
 * (already written by the runner on every live run). `extractRecordedMessages` walks such a log and
 * pulls the assistant turns in order, so the fixture is just "a prior session log + the messages
 * the LLM produced". No new on-disk format.
 */

import type { ProviderMessage, ToolCall } from "../types.js"
import type { SessionEvent } from "./session-log.js"

/**
 * Extract the ordered list of assistant Messages from a recorded session log.
 *
 * Walks `llm_completed` events (which is what the runner appends for every LLM call) and produces
 * one ProviderMessage per event. Pass the result directly to `new ReplayProvider(messages)`.
 *
 * Accepts both camelCase and snake_case tool-call wire shapes. Token usage remains session
 * evidence; the returned public ProviderMessage mirror carries no token projection.
 *
 * @param events Session events, in original order. Accepts both `{ event, seq }` (the shape
 *               `SessionLog.read()` returns) and a bare `SessionEvent[]`.
 */
export function extractRecordedMessages(
  events: Array<{ event: SessionEvent } | SessionEvent>,
): ProviderMessage[] {
  const out: ProviderMessage[] = []
  for (const entry of events) {
    const event: SessionEvent = isWrapped(entry) ? entry.event : entry
    if (event.kind !== "llm_completed") continue
    const e = event as unknown as Record<string, unknown>
    const tcRaw = (e.toolCalls ?? e.tool_calls) as unknown[] | undefined
    out.push({
      role: "assistant",
      content: typeof e.content === "string" ? e.content : "",
      ...(Array.isArray(tcRaw) && tcRaw.length > 0
        ? { toolCalls: normalizeToolCalls(tcRaw as ToolCall[]) }
        : {}),
    })
  }
  return out
}

function isWrapped(x: unknown): x is { event: SessionEvent } {
  return !!x && typeof x === "object" && "event" in (x as Record<string, unknown>)
}

function normalizeToolCalls(tcs: ToolCall[]): ToolCall[] {
  return tcs.map(tc => ({
    id: tc.id,
    name: tc.name,
    arguments: typeof tc.arguments === "string" ? tc.arguments : JSON.stringify(tc.arguments),
  }))
}
